configdir/parser.py

```diff
@@ -37,6 +39,8 @@ def _get_config_values(directory, interpolate_keys, parent):
                 if not yaml:
                     raise TypeError("YAML not supported: {}".format(path))
                 contents = yaml.safe_load(contents)
+            elif ext == ".uri":
+                key_names_to_be_encoded.add(key_name)
```

Comment:
See `interpolate_keys.add(parent + (key_name,))`. The same is needed here.
```suggestion
                key_names_to_be_encoded.add(parent + (key_name,))
```

Reply:
It would probably be better to make this more extensible. It would not be good to have to add a new parameter any time a new setting was needed.

```python
def _uri_parameter_formattter(key, value, values):
    return quote(value)

def _get_config_values(directory, key_settings, parent):
    config = {}
    for name in os.listdir(directory):
        path = os.path.join(directory, name)
        key_name, ext = os.path.splitext(name)
        if key_name in config:
            raise KeyError(
                "Duplicate config key named `{}` with path {}".format(key_name, path)
            )
        if os.path.isdir(path):
            config[key_name] = _get_config_values(
                path, key_settings, parent + (key_name,)
            )
        else:
            settings = key_settings.setdefault(parent + (key_name,), {})
            with open(path, "rb") as f:
                contents = f.read().strip()

            interpolate = True
            if ext == ".json":
                contents = json.loads(contents)
            elif ext == ".bin":
                # Leave value as is
                interpolate = False
            elif ext == ".yaml":
                if not yaml:
                    raise TypeError("YAML not supported: {}".format(path))
                contents = yaml.safe_load(contents)
            elif ext == ".uri":
                settings["formattter"] = _uri_parameter_formattter
            else:
                contents = contents.decode("utf8")

            if interpolate:
                settings["interpolate"] = True

            config[key_name] = contents
    return config


def parse(directory):
    """
    Parse a ConfigDir

    Args:
        directory (str): Directory path to config contents.

    Returns:
        dict: Config values

    Raises:
        ConfigDirMissingError: When configuration directory is missing.
    """
    if not os.path.isdir(directory):
        raise ConfigDirMissingError("`{}` is not a directory.".format(directory))

    key_settings = {}
    config = _get_config_values(directory, key_settings, ())
    interpolate = Interpolator(config)

    for key_names, settings in key_settings.items():
        c = config
        for k in key_names[:-1]:
            c = c[k]

        if settings.get("interpolate"):
            c[key_names[-1]] = interpolate(c[key_names[-1]], formatter=settings.get("formatter"))

    return config

```